### PR TITLE
ZK-4546: Caching of EL method invocations

### PR DESCRIPTION
**Pull request by DevChu**

### CODE REVIEW

File: zel/src/org/zkoss/zel/BeanELResolver.java [hunk 1]
```diff
@@ -583,4 +532,132 @@ public void put(K key, V value) {
...
 
     }
+
+	// ZK-4546
+	private static final class CachedMethodInfo {
+		private final String _methodName;
+
+		private final Class<?>[] _paramTypes;
+
+		private int _hashCode;
```
Comment (plijyfes) [inline zel/src/org/zkoss/zel/BeanELResolver.java:542]: move into constructor

File: zel/src/org/zkoss/zel/BeanELResolver.java [hunk 2]
```diff
@@ -583,4 +532,132 @@ public void put(K key, V value) {
...
+				if (_paramTypes != null)
+					result += _paramTypes.length;
+				_hashCode = result;
+			}
+			return result;
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (obj instanceof CachedMethodInfo) {
```
Comment (plijyfes) [inline zel/src/org/zkoss/zel/BeanELResolver.java:571]: refactor

File: zel/src/org/zkoss/zel/BeanELResolver.java [hunk 3]
```diff
@@ -583,4 +532,132 @@ public void put(K key, V value) {
...
+			}
+			if (matchingMethod == null) {
+				throw new MethodNotFoundException("Unable to find method ["
+						+ methodName + "] with [" + paramCount + "] parameters");
+			}
+		}
+		if (matchingMethod != null) {
+			Map<CachedMethodInfo, Method> clzMap = METHODS_CACHE.get(clazz);
+			if (clzMap == null) {
+				clzMap = new ConcurrentHashMap<>();
```
Comment (plijyfes) [inline zel/src/org/zkoss/zel/BeanELResolver.java:656]: check concurrent

**Review by plijyfes — CHANGES_REQUESTED** [review]

**Review by plijyfes — APPROVED** [review]

File: zel/src/org/zkoss/zel/BeanELResolver.java [hunk 4]
```diff
@@ -583,4 +532,124 @@ public void put(K key, V value) {
...
+					}
+				}
+			}
+			if (matchingMethod == null) {
+				throw new MethodNotFoundException("Unable to find method ["
+						+ methodName + "] with [" + paramCount + "] parameters");
+			}
+		}
+		if (matchingMethod != null) {
+			Map<CachedMethodInfo, Method> clzMap = METHODS_CACHE.get(clazz);
```
Comment (scribetw) [inline zel/src/org/zkoss/zel/BeanELResolver.java:646]: Use other Lock other than synchorized

**Review by scribetw — CHANGES_REQUESTED** [review]

File: zel/src/org/zkoss/zel/impl/util/ReflectionUtil.java [hunk 5]
```diff
@@ -49,8 +49,9 @@
             long.class, short.class, Void.TYPE };
 
 	//for reflection cache
+    private static final int DEFAULT_CAPACITY = 16;
```
Comment (plijyfes) [inline zel/src/org/zkoss/zel/impl/util/ReflectionUtil.java:52]: cacheSize getSystem

File: zel/src/org/zkoss/zel/impl/util/ReflectionUtil.java [hunk 6]
```diff
@@ -560,11 +561,7 @@ public int compareTo(MatchResult o) {
 		if (resultList != null && !resultList.isEmpty())
 			result = resultList.toArray(new Method[resultList.size()]);
 
-		Map<String, Method[]> classMap = SETTERS_METHODS_CACHE.get(cls);
-		if (classMap == null) {
-			classMap = new ConcurrentHashMap<String, Method[]>();
-			SETTERS_METHODS_CACHE.put(cls, classMap);
-		}
+		Map<String, Method[]> classMap = SETTERS_METHODS_CACHE.computeIfAbsent(cls, k -> new ConcurrentHashMap<>());
 		classMap.put(propertyName, result);
```
Comment (plijyfes) [inline zel/src/org/zkoss/zel/impl/util/ReflectionUtil.java:565]: one line

File: zk/src/org/zkoss/zk/ui/util/ConventionWire.java [hunk 7]
```diff
@@ -290,14 +290,16 @@ private void injectFieldByName(Object arg, Class tgtcls, Class parmcls, String f
...
-			injectField(arg, parmcls, fd);
-		} catch (NoSuchFieldException e) {
-			//ignore
+			Field fd = _injectedFieldCache.computeIfAbsent(cachedKey, k -> {
+				try {
+					return Classes.getAnyField(tgtcls, fdname);
+				} catch (NoSuchFieldException e) {
+					//ignore
+				}
+				return null;
```
Comment (plijyfes) [inline zk/src/org/zkoss/zk/ui/util/ConventionWire.java:299]: null value is invalid

**Review by plijyfes — CHANGES_REQUESTED** [review]

**Review by scribetw — APPROVED** [review]

**Review by scribetw — CHANGES_REQUESTED** [review]
In conventionWire, cache the "NotFound" result to prevent finding again

File: zk/src/org/zkoss/zk/ui/util/ConventionWire.java [hunk 8]
```diff
@@ -286,18 +292,28 @@ private void injectByName(Object arg, String fdname, boolean fieldOnly) {
...
-			injectField(arg, parmcls, fd);
-		} catch (NoSuchFieldException e) {
-			//ignore
+			Field fd = _injectedFieldCache.computeIfAbsent(cachedKey, k -> {
+				try {
+					return Classes.getAnyField(tgtcls, fdname);
+				} catch (NoSuchFieldException e) {
+					//ignore
+				}
+				return NotFound.class.getFields()[0];
```
Comment (scribetw) [inline zk/src/org/zkoss/zk/ui/util/ConventionWire.java:313]: private static final Field EMPTY = NotFound.class.getFields()[0];

**Review by scribetw — APPROVED** [review]

**Comment by scribetw** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I just pushed up the changes, enjoy!

**Pull request merged**